### ISSUE TITLE
Using Owned to use structs in-place instead of copying the data out

### ISSUE THREAD

**Issue by polyfractal** [dwrensha/capnproto-rust#50]
I admit defeat :)  Continuing this question from IRC...I've been unable piece together how to appropriately use Owned.  I'm not familiar enough with capnp or Rust lifetimes to work out how the various pieces are used, and there aren't many examples in the three code repositories that use Owned.

My question on IRC: 
> any capnp tips to dealing with ownership? E.g. i want to serialize a timestamp, but to "rebuild" the timestamp on the other end, Chrono needs to take ownership of the u64 from capnp. I keep running into this with anything more advanced than simple primitives

And @dwrensha's response:

>capnp::traits::Owned might help: https://github.com/dwrensha/capnproto-rust/blob/f24ab9e95f1467ff023dd3b8a51495aa1637fa4d/src/traits.rs#L41-L56

>I think I eventually want to switch message::Reader and message::Builder to be parameterized by a `T where T : for<'a> Owned<'a>`.

>imagine that message::Reader is paramterized by foo::Owned. Then we would have `got_root<'a>(&'a self) -> foo::Reader<'a>`. The point is that the container doesn't need to worry about lifetimes; we can wait until a self-borrowing accessor method to commit to a lifetime.

>and `get_root<'a>(&'a self) -> <T as capnp::traits::Owned<'a>>::Reader`

Do you happen to have a small example handy? This seems like a critical part of using Capnp correctly.  I'd be happy to contribute back documentation once I figure out how the various pieces work together :)

**Comment by dwrensha** [comment]
This is the kind of thing I had in mind:

```rust
    pub struct OwnedReader<T> where T : for<'a> ::capnp::traits::Owned<'a> {
        marker: ::std::marker::PhantomData<T>,
        message: ::capnp::message::Reader<::capnp::serialize::OwnedSegments>,
    }

    impl <T> OwnedReader<T> where T: for <'a> ::capnp::traits::Owned<'a> {
        pub fn get<'a> (&'a self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader >
        {
            self.message.get_root()
        }
    }
```

This lets you pass around an owned message without losing type information.

If the data you care about is not at the root of a message, then probably the easiest thing to do would be to copy the data into the root of a new message and then use the above pattern directly.

**Comment by vks** [comment]
How would you implement `fn get<'a>(&'a self)` for an `OwnedBuilder`?

If I have

```rust
    struct OwnedBuilder<T> where T: for<'a> ::capnp::traits::Owned<'a> {
        marker: ::std::marker::PhantomData<T>,
        message: ::capnp::message::Builder<::capnp::message::HeapAllocator>,
    }
```

then implementing `get_mut` is straightforward:

```rust
    impl<T> OwnedBuilder<T> where T: for<'a> ::capnp::traits::Owned<'a> {
        fn get_mut<'a>(&'a mut self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Builder>
        {
            self.message.get_root()
        }
    }
```

Is it possible to implement

```rust
        fn get<'a>(&'a self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader>
```

as well?

**Comment by dwrensha** [comment]
Unfortunately, that's not possible yet. One way that we might get it to work would be to define a new trait

```rust
pub trait AsReader<Reader> {
   fn as_reader(self) -> Reader;
}
```

and we would and then adjust the definition of `Owned`:

```rust
pub trait Owned<'a> {
    type Reader: FromPointerReader<'a> + SetPointerBuilder<Self::Builder>;
    type Builder: FromPointerBuilder<'a> + AsReader<Self::Reader>;
}
```

**Comment by vks** [comment]
To then get a builder again, we would need an `AsBuilder` trait, right?

**Comment by dwrensha** [comment]
No, you wouldn't be able to go from a `Reader` to a `Builder`. That would be roughly equivalent to going from a `&Foo` to a `&mut Foo`.

**Comment by dwrensha** [comment]
Note also that the generated `Builder<'a>` structs have a method `borrow_as_reader()`:

```rust
impl <'a> Reader<'a> {
  pub fn borrow_as_reader<'b>(&'b self) -> Reader<'b>  {...}
}
```

It seems like it would be useful to put this in a trait too, but it seems like we might need higher-kinded types to do so.

Like, we kind of want to define
```rust 
pub trait Owned<'a> {
    type Reader: FromPointerReader<'a> + SetPointerBuilder<Self::Builder>;
    type Builder: FromPointerBuilder<'a> + AsReader<Self::Reader> + BorrowAsReader<Self::Reader>;
}
```

But what would the definition of `BorrowAsReader` be?

```rust
pub trait BorrowAsReader<Reader> {
   fn as_reader(&'a self) -> Reader<'a>;
}
```
... this doesn't quite work because of the lifetime parameter on `Reader`.

**Comment by dwrensha** [comment]
As of https://github.com/dwrensha/capnproto-rust/commit/9b4167db7075a989b446afbb4f85049f4cc05024, which introduced `get_root_as_reader()`, we now have a satisfactory answer to https://github.com/dwrensha/capnproto-rust/issues/50#issuecomment-163057634

> How would you implement `fn get<'a>(&'a self)` for an `OwnedBuilder`?

```rust
fn get<'a>(&'a self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader> {
   self.message.get_root_as_reader()
}
```

**Comment by bspeice** [comment]
I'm trying to get an owned reference so I can send this across a thread, and I've copied the above code as follows:

```rust
pub struct OwnedReader<T> where T : for<'a> ::capnp::traits::Owned<'a> {
    marker: ::std::marker::PhantomData<T>,
    message: ::capnp::message::Reader<::capnp::serialize::OwnedSegments>,
}

impl <T> OwnedReader<T> where T: for <'a> ::capnp::traits::Owned<'a> {
    pub fn get<'a> (&'a self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader >
    {
        self.message.get_root_as_reader()
    }
}
```

And when I try to use a reader like

```rust
pub enum Archivable<'a> {

    MarketData(OwnedReader<market_data::Reader<'a>>)
}
```

I get the compile-time exception of:

```
the trait `for<'a> capnp::traits::Owned<'a>` is not implemented for `market_data_capnp::market_data::Reader<'a>`
```

Am I approaching this incorrectly? My current workaround is to just send a struct containing the message content as `Vec<u8>` and empty enum, and construct the Reader on the other side.

**Comment by dwrensha** [comment]
Does it work like this?

```rust
pub enum Archivable {
    MarketData(OwnedReader<market_data::Owned>)
}
```

**Comment by bspeice** [comment]
Yes after a slight modification to `OwnedReader`:

```rust
pub struct OwnedReader<T> where T : for<'a> ::capnp::traits::Owned<'a> {
    marker: ::std::marker::PhantomData<T>,
    message: ::capnp::message::Reader<::capnp::serialize::OwnedSegments>,
}

impl <T> OwnedReader<T> where T: for <'a> ::capnp::traits::Owned<'a> {
    pub fn get<'a> (&'a self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader >
    {
        **self.message.get_root()**
    }
}
```

Final question though: How do I create an `OwnedReader<market_data::Owned>`? Given that `market_data::Owned` is an empty struct, I'm not sure where the `Vec<u8>` is actually owned at. Thanks for the help!

**Comment by dwrensha** [comment]
Something like this should work:

```rust
impl <T> OwnedReader<T> where T : for<'a> ::capnp::traits::Owned<'a> {
    pub fn new(message: ::capnp::message::Reader<::capnp::serialize::OwnedSegments>) -> Self {
        OwnedReader {
            marker: ::std::marker::PhantomData,
            message: message,
        }
    }
}
```

can be called like this:
```rust
OwnedReader<market_data::Owned>::new(message)
```

**Comment by bspeice** [comment]
Brilliant. Got this working using standard `::capnp::serialize::read_message()`, greatly appreciate the help. Any chance this could get added to the default library? I'd be happy to write up something on how to use it, and I know this would be helpful for #23 as well.

**Comment by dwrensha** [comment]
Yeah, enough people have asked about this that it would probably make sense to add it to the library.  Seems like it would belong in the `capnp::message` module. I would lean towards making the `OwnedSegments` part generic too, like:

```rust
pub struct OwnedReader<S, T>
  where S: ::capnp::message::ReaderSegments,
        T : for<'a> ::capnp::traits::Owned<'a>
{
    marker: ::std::marker::PhantomData<T>,
    message: ::capnp::message::Reader<S>,
}
```

**Comment by bspeice** [comment]
Working on tests for recently merged code. Is there a way of creating `address_book::Owned` from static information (i.e. like in the example)? I'm currently using `serialize::read_message` to get back a `Reader<OwnedSegments>`.

**Comment by tv42** [comment]
I think the easiest path is BufferSegments, see https://github.com/capnproto/capnproto-rust/issues/259#issuecomment-1950380081 -- this may be new since the last comment from 2018.
